Openshift security best practices (#430)
UKCloud/documentation: articles/openshift/oshift-ref-security-best-practices.md

## Service accounts

You should use service accounts for all API access, which should be accomplished by an automated process; never use a regular user's credentials. Usually a service account's username is the same as the project it is created for, followed by its name:

```
system:serviceaccount:<project>:<name>
```

For example:

```
system:serviceaccount:continuous-integration:jenkins
```

You can read more about service accounts [here](https://docs.openshift.com/container-platform/3.11/dev_guide/service_accounts.html).

## Creating and deploying vulnerable container images

Just like traditional software packages, containers also run software, even if they are running that software within a self contained environment of a container. These containers can be built and run using vulnerable software packages/libraries. It is therefore considered good practice to do the following to ensure that you are not introducing vulnerable containers into your clusters.

>  1. Only use base images from trusted sources, i.e Red Hat's container registry. RedHats containers on RedHats container registry are scanned for vulnerabilities and given a security rating. Docker Hub also provides some certified containers which are built using best practices, tested and validated against the Docker Enterprise Edition platform and APIs, pass Dockers security requirements, and are collaboratively supported by Docker and the firms whos container it is. Users can also choose to build base containers from scratch using only the software packages they need for their workload. Although building containers from scratch and using just enough of the software packages you need to run for your workload does not garantee that your base image is going to be free of all possible vulnerabilities as it is possible that some of those software packages could contain vulnerabilities.

Reviewer: ```suggestion
 1. Only use base images from trusted sources, i.e Red Hat's container registry. Red Hats containers on Red Hats container registry are scanned for vulnerabilities and given a security rating. Docker Hub also provides some certified containers which are built using best practices, tested and validated against the Docker Enterprise Edition platform and APIs, pass Dockers security requirements, and are collaboratively supported by Docker and the company responsible for the container. Users can also choose to build base containers from scratch using only the software packages they need for their workload. Although building containers from scratch and using just enough of the software packages you need to run for your workload does not guarantee that your base image is going to be free of all possible vulnerabilities as it is possible that some of those software packages could contain vulnerabilities.
```

>  2. Scan your containers with container scanning software as part of your build or deployment process. There are several tools available to scan containers. These tools give you a security rating on your final container once it's been created from a secure base container. There are a few such tools:- 

Reviewer: ```suggestion
 2. Scan your containers with container vulnerability scanning software as part of your build or deployment process. There are several tools available to scan containers. These tools give you a security rating on your final container once it's been created from a secure base container. There are a number of tools available to do this, the following is a small sample we're aware of:- 
```

This article sets out some best practices to follow to minimise the attack vector to a UKCloud for OpenShift cluster and its workloads.

### Intended audience

This article is intended for anyone who is using UKCloud for OpenShift. It highlights some of the best practices to keep OpenShift workloads secure.

## Role Based Access Control

Within OpenShift, you can grant various levels of access to users, which can be cluster-wide or limited to a particular OpenShift object. You can associate users and groups with multiple roles at the same time.

You can view a pre-defined list of cluster roles and local roles [here](https://docs.openshift.com/container-platform/3.11/architecture/additional_concepts/authorization.html#roles).

 - Users with the cluster-admin default cluster role bound cluster-wide can perform any action on any resource.

 - Users with the admin default cluster role bound locally can manage roles and bindings in that project.

> It is good practice to follow the [Principle of Least Access](https://en.wikipedia.org/wiki/Principle_of_least_privilege) with OpenShift access. For information about creating new OpenShift users, see [*How to create new UKCloud for OpenShift users in the UKCloud Portal*](oshift-how-create-users.md). This article also explains how to enable two-factor authentication (2FA) when logging in to OpenShift. Users are also encouraged to change their passwords regularly, which is possible to do from the UKCloud Portal.

Reviewer 2: ```suggestion
It is good practice to follow the [Principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege) with OpenShift access. For information about creating new OpenShift users, see [*How to create new UKCloud for OpenShift users in the UKCloud Portal*](oshift-how-create-users.md). This article also explains how to enable two-factor authentication (2FA) when logging in to OpenShift. Users are also encouraged to change their passwords regularly, which is possible to do from the UKCloud Portal.
```

You should ensure that OpenShift account credentials are not shared between multiple users. Instead, you should create individual user accounts for each user with the correct role binding. It is also the responsibility of each individual user to ensure the physical security of the terminal from which they access OpenShift.

## Service accounts

> You should use service accounts for all API access, which should be accomplished by an automated process; never use an regular user's credentials. Usually a service account's username is the same as the project it is created for, followed by its name:

Reviewer 2: ```suggestion
You should use service accounts for all API access, which should be accomplished by an automated process; never use a regular user's credentials. Usually a service account's username is the same as the project it is created for, followed by its name:
```

```bash
system:serviceaccount:<project>:<name>. 
```

For example:

```bash
system:serviceaccount:continuous-integration:jenkins
```

You can read more about service accounts [here](https://docs.openshift.com/container-platform/3.11/dev_guide/service_accounts.html).

## Creating and deploying vulnerable container images

> There have been many incidents in recent history whereby containers that contain severe vulnerabilities have been deployed into production clusters. To avoid this from happening, you should use base images from trusted sources, that is, from Red Hat's container registry. These containers are scanned for vulnerabilities on a daily basis and given a security rating based on known security risks at that moment in time. Docker Hub also provides some certified containers. Users can also choose to build their own base containers from scratch. However, containers built from scratch could still contain vulnerable packages. Therefore, it is good practice to set up an automated image scanning process, as part of your container building and deploying process, which can scan the built image and provide an image security rating. It is also possible to use image scanning tools to certify your own images before they are deployed so that only images with an acceptable security rating set by your OpenShift administrators are deployed.

Reviewer 2: ```suggestion
There have been many incidents in recent history whereby containers that contain severe vulnerabilities have been deployed into production clusters. To avoid this from happening, you should use base images from trusted sources, that is, from Red Hat's container registry. These containers are scanned for vulnerabilities on a daily basis and given a security rating based on known security risks at that moment in time. Docker Hub also provides some certified containers. Users can also choose to build their own base containers from scratch. However, containers built from scratch could still contain vulnerable packages. Therefore, it is good practice to set up an automated image scanning process, as part of your container building and deployment process, which can scan the built image and provide an image security rating. It is also possible to use image scanning tools to certify your own images before they are deployed so that only images with an acceptable security rating set by your OpenShift administrators are deployed.
```

> There have been many incidents in recent history whereby containers that contain severe vulnerabilities have been deployed into production clusters. To avoid this from happening, you should use base images from trusted sources, that is, from Red Hat's container registry. These containers are scanned for vulnerabilities on a daily basis and given a security rating based on known security risks at that moment in time. Docker Hub also provides some certified containers. Users can also choose to build their own base containers from scratch. However, containers built from scratch could still contain vulnerable packages. Therefore, it is good practice to set up an automated image scanning process, as part of your container building and deployment process, which can scan the built image and provide an image security rating. It is also possible to use image scanning tools to certify your own images before they are deployed so that only images with an acceptable security rating set by your OpenShift administrators are deployed.

Reviewer: I'm not sure I like this paragraph. It sounds too much to me like we've seen many incidents of this, as if we've made this mistake ourselves. If we're going to be mentioning this here we should at least link to news articles of this happening elsewhere. Otherwise I'd prefer we just talk about it being a potential attack vector rather than be so conversational about it.

Additionally we should mention using containers from the Red Hat container registry as base images as they have been scanned for vulnerabilities and the score they achieve is visible on the container registry site.

Some container image scanning tools include:

  -  OpenSCAP-docker

  -  Black Duck Hub

  -  JFrog Xray

  -  Twistlock

  -  Image-inspector

  -  Clair

> You should also apply the Principle of least Access with containers too. You should never run containers as the root user; instead always run as a user with just enough privileges to run the required container workload.

Reviewer 2: ```suggestion
You should also apply the Principle of least privilege with containers too. You should never run containers as the root user; instead always run as a user with just enough privileges to run the required container workload.
```

> You should also apply the Principle of least privilege with containers too. If it can be avoided you should never run containers as the root user; instead always run as a user with just enough privileges to run the required container workload.

Reviewer: We should point out here how containers don't run as root by default and that you have to take steps to make them run as privileged, and that this will often cause containers that try to run as root to fail. We can add another set of docs around running privileged containers later to this section.

## Secrets

> Sensitive information, such as passwords, SSH keys, tokens and certificates, are examples of information that could be considered secret. Such secrets should never be stored in any place in readable format (for example, in source code, GitHub repositories, and so on). It is imperative that when secrets are required in OpenShift, they are set up as OpenShift Secrets. This allows the secrets to be encrypted within OpenShift and accessed only by the relevant containers when they are needed.

Reviewer: I think we should expand this section and point out the limitations of secrets (e.g. base64 encoded, potentially accessible from any container on the same node). Whilst its a tool to get started with there are better secret options out there like vault that we could be recommending.

## Limiting resources

You should always set up containers with resource limits. In the event of a denial of service attack, or even just a burst of very high load on a service, these resource limits will prevent pods from using all available resources in the node they are running on. You can read more about setting resource quotas and limits [here](https://docs.openshift.com/container-platform/3.11/dev_guide/compute_resources.html).

## Privileged ports

> TCP/IP ports below port 1024 are considered to be privileged ports and you should not use them to front any service that is deployed into OpenShift.

Reviewer: This needs to be clearer - does this mean we can't expose port 80 on a container within the pod network? Thats obviously not the case, but I don't think its very clear what we mean here. Unless we do mean that of course - but that's a pretty common scenario so what's the alternative.

## Route security

> Within OpenShift, you can expose your services using routes. Creating a route to a service enables your users to access the service through a URL on a browser. By default, the URLs that are created are exposed on HTTP port 80 without any SSL certificate. You can secure these routes with SSL encryption using one of three SSL termination methods:

Reviewer: Is this true? I think if you're exposing port 443 then you'll find that port 80 isn't the default for the route.

 - Edge - SSL termination is done on the HAProxy router from which point onwards the traffic is un-encrypted

 - Passthrough termination - There is no SSL decryption on the routers; instead connections are passed straight through the routers to the OpenShift applications, which are then responsible for SSL certificate termination and traffic decryption.

 - Re-Encrypt - This process uses two certificates: the first certificate is held on the HAProxy router and the second in the pods. This secures traffic all the way down to the application.

> It is possible to further secure routes by applying IP whitelisting on routes. We have documented the process of IP whitelisting [here](oshift-how-restrict-access-to-openshift-routes-by-ip-address.md).

Reviewer: Can we break this out into a separate section about whitelisting and describe where the traffic gets dropped. E.g. with route whitelisting its at the router, but inside the cluster whereas with security groups it never gets to the cluster itself and is dropped by the external platform.

> UKCloud can whitelist individual IP addresses or entire subnets to access the OpenShift console and metrics page. It is also possible to do the same for the services that you create and deploy in OpenShift, even if those services are exposed to various Government Community Networks with the use of security groups. These restrictions can be applied at the time the cluster is deployed or can be done later via a Service Request to the UKCloud OpenShift team.

Reviewer: Make this a bit clearer, we can whitelist the control plane (which includes the following infrastructure components) and we can also whitelist the dataplanes on any network that is attached. Be clear that this blocks access to all dataplane routes from the given network but is an effective security control.

## Software-defined networking (SDN)

OpenShift uses a software-defined networking (SDN) approach to provide cluster-wide networking that enables communication between pods across the OpenShift cluster. This Container Networking Interface (CNI) runs in the OpenShift SDN project and creates an overlay network using the Open vSwitch (OVS) networking plugin.

> However, from February 2019, UKCloud will be deploying OpenShift clusters with the ovs-networkpolicy plugin. This plugin allows project administrators to configure their own isolation policies using [NetworkPolicy objects](https://docs.openshift.com/container-platform/3.6/admin_guide/managing_networking.html#admin-guide-networking-networkpolicy).

Reviewer 2: ```suggestion
However, from February 2019, UKCloud will be deploying OpenShift clusters with the ovs-networkpolicy plugin. This plugin allows project administrators to configure their own isolation policies using [NetworkPolicy objects](https://docs.openshift.com/container-platform/3.6/admin_guide/managing_networking.html#admin-guide-networking-networkpolicy). You can find an article regarding NetworkPolicy [here](oshift-netpol.md).
```

> However, from February 2019, UKCloud will be deploying OpenShift clusters with the ovs-networkpolicy plugin. This plugin allows project administrators to configure their own isolation policies using [NetworkPolicy objects](https://docs.openshift.com/container-platform/3.11/admin_guide/managing_networking.html#admin-guide-networking-networkpolicy). You can find an article regarding NetworkPolicy [here](oshift-netpol.md).

Reviewer: This reads to me like we're not going to be using ovs in the network policy SDN, but I think it still does. Perhaps removing the however will fix that. I think we should remove the date reference as well and just state in earlier versions we deployed with ovs-multitenant, and now we deploy with ovs-networkpolicy and view this article to find out more, but basically one has more granular control but is easier to open access inadvertently.

> However, from February 2019, UKCloud will be deploying OpenShift clusters with the ovs-networkpolicy plugin. This plugin allows project administrators to configure their own isolation policies using [NetworkPolicy objects](https://docs.openshift.com/container-platform/3.11/admin_guide/managing_networking.html#admin-guide-networking-networkpolicy). You can find an article regarding NetworkPolicy [here](oshift-netpol.md).

Reviewer: Also the netpol link doesn't seem to work?
